### PR TITLE
refactor(module-tools): merge libuild to module tools and redesign the hooks

### PR DESCRIPTION
**Pull request by 10Derozan**
### <samp>🤖 Generated by Copilot at c0fdf5e</samp>

This pull request updates the documentation and the dependencies for the `modern.js` project. It removes some outdated or unnecessary code, fixes some errors and typos, and adds a new option for the build config API. It also adds the `@modern-js/utils` package as a dependency to all the generator packages, which provides some common utility functions for the generator templates.

### <samp>🤖 Generated by Copilot at c0fdf5e</samp>

*  Remove unused or deprecated options and functions related to the build config and the SWC transform ([link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-a7c5f07b8103f54cc7910493cc4598a797a04f0cd3f50d351753d1e86ce5742dL147-L151), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-90a0f132c4140e37c110b40455aa3d8ad4d6a89ca7a8c8d19aa1b9a975b2d3baL295-L303), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-90a0f132c4140e37c110b40455aa3d8ad4d6a89ca7a8c8d19aa1b9a975b2d3baL316), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-90a0f132c4140e37c110b40455aa3d8ad4d6a89ca7a8c8d19aa1b9a975b2d3baL399-L415), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-c586a6521b4022a3ce9cca9155cb7658a6d5808af1138b83eb5bcd5d11e48791L295-R295), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-c586a6521b4022a3ce9cca9155cb7658a6d5808af1138b83eb5bcd5d11e48791L316), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-c586a6521b4022a3ce9cca9155cb7658a6d5808af1138b83eb5bcd5d11e48791L398-L414))
*  Add `tsconfig` option to specify the path to the tsconfig file for the build config and the dts generation ([link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-90a0f132c4140e37c110b40455aa3d8ad4d6a89ca7a8c8d19aa1b9a975b2d3baL1244-R1227), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-c586a6521b4022a3ce9cca9155cb7658a6d5808af1138b83eb5bcd5d11e48791L1243-R1226))
*  Fix typo and rendering error in the documentation for the build config ([link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-90a0f132c4140e37c110b40455aa3d8ad4d6a89ca7a8c8d19aa1b9a975b2d3baL567-R540), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-c586a6521b4022a3ce9cca9155cb7658a6d5808af1138b83eb5bcd5d11e48791L632))
*  Update documentation for the `transformLodash` option to clarify that it is enabled by default and done by Module Tools ([link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-90a0f132c4140e37c110b40455aa3d8ad4d6a89ca7a8c8d19aa1b9a975b2d3baL1222-R1195), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-c586a6521b4022a3ce9cca9155cb7658a6d5808af1138b83eb5bcd5d11e48791L1219-R1194))
*  Add `@modern-js/utils` package as a dependency to various generator packages to provide utility functions for the generator templates ([link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-e7636ad59b2d995453069533c4333a26384dcb0cdc9c43e9d535c4dab88214bdR37), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-03b429766ce926b92d87b5841676f75f7cfa39231a3c603e657c25b72a191f32R39), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-f009ade25b0c3afb4ff21d732688e68d776cbb5fdf2d10bb252f0177dce95f04R35), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-057c18d559bde0588a6ee142098c9844d9263394b5f91c17a3a8a3ef878560e7R38), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-989c0d6714d936151f926426af2fa12e8de76c604e17269da46432e6f96a4887R39), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-63d011605457243472f3716e925d18a7b2af4905a888ce50ac64cecd08ffee7fR41), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-e76e032b35378172c0a88e87eb3cbe7b1294ef328b37133c5bb154b2b32997beR39), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-2db4756ff0b0326b4df138ec6ff6f777dedd4dbd679014abe5a816e3a46e5a91R41), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-25e249731284215346d9a36be7c5811b0148643631587b8fc916c0b4450605f6R42), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-56b420a0293f65602db387e571f26e6a0189f2f7564db567640115db8a432c4eR37), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-0af5eaf01d57771bcaa5fdb47abfd6b467fd22d3720c484aab56be8e96c01c56R38), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-117b60b553ae1b77bf493590e04a7e46bf3e172c329d86af4e613f536d10f805R38), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-621200cbb00df5bc6632991906ab362668482b23a0ab1273747fec79b4fc0281R39), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-4c7391af439202832ecfd0e80c2d02cfa01f57933a50700765d8c27250b60c9fR38), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-4c1410eb4f0981198772d57977ec36f1ecb2be20bbcccb40ef9de676c6ef69ffR39), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-8a9b9baa5be33de58c2304e7f03c2cb5f772c3390df19d8c8b209a2a4d71a515R37), [link](https://github.com/web-infra-dev/modern.js/pull/4651/files?diff=unified&w=0#diff-5c57cc583353d58d9e5af6ba35c1e816ea04adbe59d59046bb4cdd88d4064bf3R39))

## Checklist

- [ ] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/cli/plugin-tailwind/src/cli.ts [hunk 1]
```diff
@@ -144,11 +144,6 @@ export const tailwindcssPlugin = (
 
         return config;
       },
-
-      modifyLibuild(config, next) {
-        config.transformCache = false;
```
Comment (10Derozan) [inline packages/cli/plugin-tailwind/src/cli.ts:149]: have not fix!

File: packages/document/module-doc/docs/en/api/config/build-config.mdx [hunk 2]
```diff
@@ -292,15 +292,6 @@ To prevent excessive global replacement substitution, it is recommended that the
...
 :::
 
-{/* ## disableSwcTransform
-
-Starting with version 2.16.0, SWC Transform is enabled by default for code transformation. If you want to disable this feature, you can use this configuration. Only esbuild Transform is used in this case.
-
-The use of SWC Transform can reduce the impact of auxiliary functions on the volume of the constructed product.
-
-* **Type**: `boolean`
-* **Default**: `false` */}
```
Comment (10Derozan) [inline packages/document/module-doc/docs/en/api/config/build-config.mdx:302]: add more doc

File: .changeset/soft-experts-roll.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+---
+'@modern-js/module-tools': minor
+---
+
+refactor(module-tools):
+
+1. merge libuild to module tools, add buildConfig.hooks to support load, transform and renderChunk
+2. breaking change. remove buildConfig.dts.tsconfigPath, use buildConfig.tsconfig instead, and improve this feature.
```
Comment (chenjiahan) [inline .changeset/soft-experts-roll.md:8]: Add compat logic for previous config

File: packages/document/module-doc/docs/en/api/config/build-config.mdx [hunk 4]
```diff
@@ -1216,10 +1226,12 @@ Specifies whether to modularize the import of [lodash](https://www.npmjs.com/pac
 
 This optimization is implemented using [babel-plugin-lodash](https://www.npmjs.com/package/babel-plugin-lodash) and [swc-plugin-lodash](https://github.com/web-infra-dev/swc-plugins/tree/main/crates/plugin_lodash) under the hood.
 
+With this configuration, the code will be converted using SWC.
+
 - **Type**: `boolean`
 - **Default**: `true`
 
-This option is enabled by default, and Modern.js Module will automatically redirects the code references of `lodash` to sub-paths.
+When you enable this, Module Tools will automatically redirects the code references of `lodash` to sub-paths.
```
Comment (chenjiahan) [inline packages/document/module-doc/docs/en/api/config/build-config.mdx:1234]:
```suggestion
When you enable this, Modern.js Module will automatically redirects the code references of `lodash` to sub-paths.
```

File: packages/document/module-doc/docs/en/api/config/build-config.mdx [hunk 5]
```diff
@@ -1216,10 +1226,12 @@ Specifies whether to modularize the import of [lodash](https://www.npmjs.com/pac
 
 This optimization is implemented using [babel-plugin-lodash](https://www.npmjs.com/package/babel-plugin-lodash) and [swc-plugin-lodash](https://github.com/web-infra-dev/swc-plugins/tree/main/crates/plugin_lodash) under the hood.
 
+With this configuration, the code will be converted using SWC.
+
 - **Type**: `boolean`
 - **Default**: `true`
```
Comment (chenjiahan) [inline packages/document/module-doc/docs/en/api/config/build-config.mdx:1232]: Now false?

File: packages/document/module-doc/docs/zh/guide/advance/in-depth-about-build.md [hunk 6]
```diff
@@ -33,34 +31,54 @@ bundleless 是单文件编译模式，因此对于类型的引用和导出你需
...
+  buildConfig: {
+    input: ['src/runtime'],
+    sourceDir: 'src',
+  },
+});
+```
+
+## 使用 swc
+
+在部分场景下，Esbuild 不足以满足我们的需求，此时我们会使用 SWC 来做代码转换，主要有以下几个场景:
```
Comment (chenjiahan) [inline packages/document/module-doc/docs/zh/guide/advance/in-depth-about-build.md:68]: 正确拼写为 esbuild

File: packages/module/plugin-module-polyfill/package.json [hunk 7]
```diff
@@ -9,47 +16,40 @@
...
-    "dev": "mdn build --watch",
     "build": "mdn build",
+    "dev": "mdn build --watch",
+    "prepublishOnly": "only-allow-pnpm",
     "test": "jest --passWithNoTests"
   },
+  "dependencies": {
+    "@babel/core": "^7.21.8",
+    "@babel/plugin-syntax-jsx": "7.21.4",
+    "@babel/plugin-syntax-typescript": "7.21.4",
```
Comment (chenjiahan) [inline packages/module/plugin-module-polyfill/package.json:31]: Why downgrade the babel version

Comment (chenjiahan) [reply packages/module/plugin-module-polyfill/package.json:31]: We should use the `^7.22.5`

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**